### PR TITLE
Pytorch refactor

### PR DESCRIPTION
**Pull request by SaeedAnas**
The pytorch code should be runnable, but there might be issues with the compute loss.

### CODE REVIEW

File: Models/MCTS_torch.py [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+            action_pos = np.argmax([v for v in visit_counts])
+        else:
+            action_pos = np.random.choice(len(visit_counts), p=action_probs)
+
+        count_entropy = entropy(action_probs, base=2)
+        return action_pos, count_entropy
+
+    @staticmethod
+    def encode_action_to_str(policy_logits, mask):
+        actions = []
```
Comment (silverlight6) [inline Models/MCTS_torch.py:177]: This function changed on main a bit, when the merge happens, you will simply need to copy.

File: Models/MCTS_torch.py [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+          self.run_batch_mcts(roots_cpp, hidden_state_pool)
+          roots_distributions = roots_cpp.get_distributions()
+
+          actions = []
+          target_policy = []
+          temp = self.visit_softmax_temperature()  # controls the way actions are chosen
+          for i in range(self.NUM_ALIVE):
+              deterministic = False  # False = sample distribution, True = argmax
+              distributions = roots_distributions[i]
+              action, _ = self.select_action(distributions, temperature=temp, deterministic=deterministic)
```
Comment (silverlight6) [inline Models/MCTS_torch.py:76]: Update to main

File: Models/MuZero_torch_agent.py [hunk 3]
```diff
@@ -200,14 +248,14 @@ def __init__(self,
             max_value = contractive_mapping(max_value)
             min_value = contractive_mapping(min_value)
         if num_steps <= 0:
-            num_steps = torch.ceil(max_value) + 1 - torch.floor(min_value)
+            num_steps = np.ceil(max_value) + 1 - np.floor(min_value)
         self.min_value = min_value
         self.max_value = max_value
         self.value_range = max_value - min_value
```
Comment (silverlight6) [inline Models/MuZero_torch_agent.py:254]: Check if we are using this class at all. If not, we want to delete this extra code.

File: Models/MuZero_torch_trainer.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+        lr = config.SGD_LR_INIT * config.SGD_LR_DECAY_RATE ** (
+            train_step / config.SGD_LR_DECAY_STEPS
+        )
+
+        for param_group in self.optimizer.param_groups:
+            param_group["lr"] = lr
+
+    def train_network(self, batch, agent, train_step, summary_writer):
+        observation, history, value_mask, reward_mask, policy_mask, value, reward, policy = batch
+        # if i % config.checkpoint_interval == 0:
```
Comment (silverlight6) [inline Models/MuZero_torch_trainer.py:39]: comments obviously shouldn't be here. Change history to action_history as well

**Review by silverlight6 — APPROVED** [review]
